library/math/dual.lisp

```diff
@@ -0,0 +1,177 @@
+
+(coalton-toplevel
+
+  (define-type (Dual :t)
+    "Representation of a Dual number in the form (a + bε) where a and b are  real numbers and ε satisfie ε**2 = 0 and ε != 0.  Note: `Eq`, and `Ord` and `Hash` only work on the primal component."
+    (Dual :t :t))
+   
+  (declare primal-part (Dual :t -> :t))
+  (define (primal-part (Dual p _))
+    p)
+
+  (declare dual-part (Dual :t -> :t))
+  (define (dual-part (Dual _ d))
+    d)
+
+  (define (sq x)
+    (* x x))
+   
+  (define-instance (Eq :t => Eq (Dual :t))
+    "Note: Eq only works on the primal components."
+    (define (== (Dual a _) (Dual p _))
+      (== a p)))
+  
+  (define-instance (Num :t => Num (Dual :t)) 
+    (define (+ (Dual p1 d1) (Dual p2 d2))
+      (Dual (+ p1 p2) (+ d1 d2)))
+     
+    (define (- (Dual p1 d1) (Dual p2 d2))
+      (Dual (- p1 p2) (- d1 d2)))
+
+    (define (* (Dual p1 d1) (Dual p2 d2))
+      (Dual (* p1 p2)
+            (+ (* p1  d2) (* d1 p2))))
+     
+    (define (fromInt z)
+      (Dual (fromInt z) 0)))
+   
+  (define-instance (Reciprocable :t => Reciprocable (Dual :t))
+    (define (/ (Dual p1 d1) (Dual p2 d2))
+      (Dual (/ p1 p2)
+            (/ (- (* d1 p2)
+                  (* p1 d2))
+  	       (* p2 p2))))
+     
+    (define (reciprocal (Dual p1 d1))
+      (Dual (reciprocal p1)
+            (/ (negate d1) (* p1 p1)))))
+
+  (define-instance ((Num :t) (Trigonometric :t) (Reciprocable :t) (Radical :t) => (Trigonometric (Dual :t)))
+    (define (sin (Dual p1 d1))
+      (Dual (sin p1)
+            (* d1 (cos p1))))
+     
+    (define (cos (Dual p1 d1))
+      (Dual (cos p1)
+            (negate (* d1 (sin p1)))))
+     
+    (define (tan (Dual p1 d1))
+      (Dual (tan p1)
+            (/ d1 (sq (cos p1)))))
+     
+    (define (asin (Dual p1 d1))
+      (Dual (asin p1)
+            (/ d1 (sqrt (- 1 (sq p1))))))
+     
+    (define (acos (Dual p1 d1))
+      (Dual (acos p1)
+            (negate (/ d1 (sqrt (- 1 (sq p1)))))))
+     
+    (define (atan (Dual p1 d1))
+      (Dual (atan p1)
+            (/ d1 (+ 1 (sq p1))))))
+
+  (define-instance ((Num :t) (Exponentiable :t) (Reciprocable :t) => (Exponentiable (Dual :t)))
+    (define (exp (Dual p1 d1))
+      (Dual (exp p1)
+            (* d1 (exp p1))))
+
+    (define (pow dual1 dual2)
+      (exp (* dual2 (ln dual1))))
+     
+    (define (ln (Dual p1 d1))
+      (Dual (ln p1)
+            (/ d1 p1)))
+     
+    (define (log dual1 dual2)
+      (/ (ln dual2) (ln dual1))))
+   
+  (define-instance ((Num :t) (Radical :t) (Reciprocable :t) (Exponentiable :t) => (Radical (Dual :t)))
+    (define (nth-root n (Dual p1 d1))
+      (let ((n* (fromInt n)))
+        (Dual (nth-root n p1)
+  	      (pow (* p1 (/ d1 n*)) (- (reciprocal n*) 1)))))
+     
+    (define (sqrt (Dual p1 d1))
+      (Dual (sqrt p1)
+            (/ p1 (* 2 (sqrt p1))))))
+
+  (define-instance ((Ord :t) (Ord :t) => Ord (Dual :t))
+    "Note: Ord only works on the primal components."
```

Comment:
"works on" -> "uses"

Reply:
I would personally prefer "Dual numbers are <typeclass operator simple past tense> <on/by>" rather than "Note: <typeclass name> only works on." The current version is more succinct, but I'd rather have docstrings that read as complete sentences.